### PR TITLE
Do not clear headers when switching between tabs

### PR DESCRIPTION
**Pull request by klippx**
## Issue

When switching between tabs, the default header state is lost and replaced with empty string. 

## Demo of the bug:

https://github.com/graphql/graphiql/assets/2785359/6bc78e0d-9ad1-4c2d-9e9b-8b3b508f2f79

## Codepen to verify/reproduce the issue:

In this example we are using defaultHeaders, and when using multiple tabs, upon reload the headers are cleared when switching tabs.

https://codesandbox.io/p/sandbox/determined-brook-rp683x?file=/package.json:11,23

Note: If you edit the headers before switching tabs, they are not cleared. I am guessing that they are then instead stored in the internal state.

### CODE REVIEW

**Comment by acao** [comment]
good find, thank you!

**Comment by klippx** [comment]
@acao Would it be OK to merge?

File: packages/graphiql-react/src/editor/tabs.ts [hunk 1]
```diff
@@ -278,7 +278,7 @@ export function useSetEditorValues({
     }) => {
       queryEditor?.setValue(query ?? '');
       variableEditor?.setValue(variables ?? '');
-      headerEditor?.setValue(headers ?? '');
+      headerEditor?.setValue(headers ?? headerEditor?.getValue() ?? '');
```
[the remark this answers is not in the document]
Comment (klippx) [reply packages/graphiql-react/src/editor/tabs.ts]: That sounds good.

**Comment by klippx** [comment]
I pushed another commit since it would just use whatever you have in the editor and not overwrite it, using default headers is a better choice when switching tabs. 

Before: 
- Default headers only apply when creating a NEW tab
After:
- When switching tab, apply headers || defaultHeaders

**Comment by klippx** [comment]
I have two PRs open that initially seem well received, how come we don't move forward and merge if it seems like a good change? @acao

**Comment by acao** [comment]
@klippx sorry I have been super busy lately and there was something else I needed to check, I cannot remember what held me up last time but I think this should be good to go, perhaps @thomasheyenbrock if you have a chance to look and think this is good to go feel free to merge

**Comment by klippx** [comment]
Sorry, I pushed a commit that was rebased on main, but removed it again since it contained weird stuff, will stop interfering unless you ask me to do something! 🙏

**Comment by klippx** [comment]
@thomasheyenbrock can you take a look at this as well? 🙏

**Comment by klippx** [comment]
Is this broken in main? 
```
Summary of all failing tests
 FAIL  packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts (17.46 s)
  ● MessageProcessor with config › caches files and schema with a URL config

    ENOENT, no such file or directory '/tmp/graphql-language-service/test/projects/default/generated-schema.graphql'
```

Seems like a bad file pointer is checked in:

```
aa6dbbb45 (Rikki Schulte 2024-05-28 08:27:39 +0200  59) const genSchemaPath =
aa6dbbb45 (Rikki Schulte 2024-05-28 08:27:39 +0200  60)   '/tmp/graphql-language-service/test/projects/default/generated-schema.graphql';
```

**Comment by klippx** [comment]
@dimaMachina 👀 want to take a look?

File: packages/graphiql-react/src/editor/tabs.ts [hunk 2]
```diff
@@ -278,10 +280,12 @@ export function useSetEditorValues({
     }) => {
       queryEditor?.setValue(query ?? '');
       variableEditor?.setValue(variables ?? '');
-      headerEditor?.setValue(headers ?? '');
+      if (typeof headers === 'string' || typeof defaultHeaders === 'string') {
+        headerEditor?.setValue(headers ?? defaultHeaders ?? '');
```
Comment (dimaMachina) [inline packages/graphiql-react/src/editor/tabs.ts:284]:
```suggestion
        headerEditor?.setValue(headers ?? defaultHeaders);
```
not needed either `headers` or `defaultHeaders` will be strings

File: packages/graphiql/cypress/e2e/headers.cy.ts [hunk 3]
```diff
@@ -22,5 +22,17 @@ describe('Headers', () => {
...
+    it('do not clear `defaultHeaders` when switching between tabs upon reload', () => {
+      cy.visit(`/?query={test}&defaultHeaders=${DEFAULT_HEADERS}`);
+      cy.assertHasValues({ query: '{test}', headersString: DEFAULT_HEADERS });
+      cy.get('.graphiql-tab-add').click();
+      cy.assertHasValues({ query: '', headersString: DEFAULT_HEADERS });
+      cy.visit('/');
+      cy.assertHasValues({ query: '', headersString: DEFAULT_HEADERS });
+      cy.get('.graphiql-tab-button').eq(0).click();
+      cy.assertHasValues({ query: '{test}', headersString: DEFAULT_HEADERS });
+    });
```
Comment (dimaMachina) [inline packages/graphiql/cypress/e2e/headers.cy.ts:36]:
```suggestion

    // https://github.com/graphql/graphiql/pull/3552
    it('do not clear `defaultHeaders` when switching between tabs upon reload', () => {
      cy.visit(`/?query={test}&defaultHeaders=${DEFAULT_HEADERS}`);
      cy.assertHasValues({ query: '{test}', headersString: DEFAULT_HEADERS });
      cy.get('.graphiql-tab-add').click();
      cy.assertHasValues({ query: '', headersString: DEFAULT_HEADERS });
      cy.visit('/');
      cy.assertHasValues({ query: '', headersString: DEFAULT_HEADERS });
      cy.get('.graphiql-tab-button').eq(0).click();
      cy.assertHasValues({ query: '{test}', headersString: DEFAULT_HEADERS });
    });
```

Comment (dimaMachina) [inline packages/graphiql/cypress/e2e/headers.cy.ts:36]:
```suggestion
```

File: .changeset/silly-hornets-kiss.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@graphiql/react': patch
+---
+
+Do not clear default headers when switching between tabs
```
Comment (dimaMachina) [inline .changeset/silly-hornets-kiss.md:5]:
```suggestion
do not clear `defaultHeaders` when switching between tabs upon reload
```

**Review by dimaMachina — APPROVED** [review]

**Pull request merged**